### DISCUSSION TITLE
How is Jolt with networked games?

### DISCUSSION THREAD

**Discussion by ValorZard** [jrouwe/JoltPhysics#768] [Q&A]
I'm curious if anyone has tested Jolt with online multiplayer games. Most of the examples I've seen using Jolt involve single-player games like Horizon Forbidden West, and I'm curious if Jolt has been battle-tested with online multiplayer. Are there any special design considerations to keep in mind when using it for multiplayer? Any footguns to avoid?

**Comment by jrouwe** [comment]
I saw this X post the other day:

https://x.com/themirrorgdp/status/1718019599361323023?s=20

And I think [X4 Foundations](https://store.steampowered.com/app/392160/X4_Foundations/), which uses Jolt, also supports [some form of multiplayer](https://www.gamewatcher.com/news/x4-foundations-online-multiplayer-release).

Several people have been asking questions relating to online projects:

https://github.com/jrouwe/JoltPhysics/discussions/611
https://github.com/jrouwe/JoltPhysics/discussions/632

There are also some tips in the documentation, see section [Rolling Back a Simulation](https://jrouwe.github.io/JoltPhysics/) and the section preceding it.

**Reply by Kluskey** [reply]
Ahoy! Yes we're using Jolt for The Mirror. Much appreciated for the awesome work with Jolt @jrouwe!

Over time, we'll be exposing a ton of functionality to developers on our platform so you'll be able to a get a "feel" for networked Jolt if you'd like to try it out. Our next release (with Jolt) is coming this month, although it's still an alpha version: https://themirror.space

Considerations: It's definitely been a heavy technical lift; we planned for 1-2 months and we're at ~6 months now for the full implementation. Strong prior knowledge of networking and physics is needed.

**Reply by jrouwe** [reply]
Thank you for using it! It's fun to see that there is so much interest in Jolt. I'll add The Mirror to the list of projects using Jolt!

**Reply by TIHan** [reply]
I only dabble in game development on the side, but I did get Jolt to work well with rollback and corrections under a high amount of simulated lag. For the rollback, I'm not using Jolt's built-in Save/Restore state functions, I'm doing it manually with `SetPositionRotationAndVelocity` and CharacterVirtual's `ExtendedUpdate` - it doesn't need to be 100% perfect determinism as the corrections fix any inconsistencies, just enough where the player wouldn't notice.

Really cool project @Kluskey and glad to see you all got it working!

**Reply by NEO97online** [reply]
@TIHan I'm currently working on something similar, and it's 90% working but I experience a lot of jitter that I haven't been able to figure out yet. It seems like the jitter happens with CharacterVirtual's stair-stepping, going down slopes, etc.

When you restore state for rollback, are you ONLY calling SetPositionRotationAndVelocity or is there more to it? Do you call RefreshContacts or anything else?

Also, how are you involving delta time in this?

Thank you.

**Reply by cuppajoeman** [reply]
@NEO97online I already discussed a similar issue to this with Jorrit a while ago here: https://github.com/jrouwe/JoltPhysics/discussions/1137. 

The issue that exists right now  is that slamming in authorative pos, vel, etc from server and manually and then calling RefreshContacts, doesn't always produce the same ground state as would be if you had just done an ExtendedUpdate to produce that state, this is explained here: 

> I think RefreshContacts should mostly fix this issue as it runs collision detection again and updates the ground state, however it is not 100% the same thing as calling ExtendedUpdate as this can also detect contacts that you touch during the update (but don't end up touching at the end of the step).

I have a hack where I simply call RefreshContacts after every single ExtendedUpdate you call, that way you can guarentee that the ground state will be the same, whether it's produced by manually updating vs slamming in the server's authorative state. This is a hack because I think ExtendedUpdate's logic for determining those attributes is probably more accurate, and additionally you're running collision detection twice.

This hack is fine for now, as it allows me reliably reproduce my movement logic on client and server, so for me this was this biggest thing to realize if you're going with manually setting the state method, instead of RestoreState.

As for delta times, when you're simulating steps both on client and server, then if you use measured deltas, you'll have the issue that the client will update with one delta time eg 0.016668 and the server might update with time 0.0166662 which are different and thus will produce different states, so instead what I do is that I have a way of handling dt's so that you pass out the perfect period eg 1/freq whenever at least that much time has passed so that the same dt's are used both on client and server, which lead me to less jitter.

After learning about the [SaveState](https://github.com/jrouwe/JoltPhysics/blob/2bce8a6bdcafdf6b1140a080ebedcfbcdba74313/Jolt/Physics/Character/CharacterVirtual.cpp#L1856) / RestoreState functions, I think it's probably cleaner and would work better to send over the StateRecorderImpl's over the network, but just haven't gotten around to it yet, but will try it some day.
